bin/pkg.ml

```diff
@@ -101,25 +101,85 @@ module Lock = struct
         (opam_package_name, opam_file))
     |> OpamPackage.Name.Map.of_list
 
+  (* Logic for choosing the lockdir path(s) to generate *)
+  let choose_lock_dir_paths ~context_name_arg ~all_contexts_arg =
+    let open Fiber.O in
+    match (context_name_arg, all_contexts_arg) with
+    | Some _, true ->
+      User_error.raise
+        [ Pp.text "--context and --all-contexts are mutually exclusive" ]
+    | context_name_opt, false -> (
+      let+ workspace = Memo.run (Workspace.workspace ()) in
+      let context_name =
+        Option.value context_name_opt ~default:Dune_engine.Context_name.default
+      in
+      let context =
+        List.find workspace.contexts ~f:(fun context ->
+            Dune_engine.Context_name.equal
+              (Workspace.Context.name context)
+              context_name)
+      in
+      match context with
+      | None ->
+        User_error.raise
+          [ Pp.textf "Unknown build context: %s"
+              (Dune_engine.Context_name.to_string context_name
+              |> String.maybe_quoted)
+          ]
+      | Some (Default { lock; _ }) ->
+        [ Option.value lock ~default:Lock_dir.default_path ]
+      | Some (Opam _) ->
+        User_error.raise
+          [ Pp.textf "Unexpected opam build context: %s"
+              (Dune_engine.Context_name.to_string context_name
+              |> String.maybe_quoted)
+          ])
+    | None, true ->
+      let+ workspace = Memo.run (Workspace.workspace ()) in
+      List.filter_map workspace.contexts ~f:(function
+        | Workspace.Context.Default { lock; _ } -> lock
+        | Opam _ -> None)
+
   let term =
     let+ (common : Common.t) = Common.term
-    and+ repo_selection = Repo_selection.term in
+    and+ repo_selection = Repo_selection.term
+    and+ context_name =
+      Common.context_arg_opt
+        ~doc:
+          "Generate the lockdir associated with this context (the default \
+           context will be used if this is omitted)"
+    and+ all_contexts =
+      Arg.(
+        value & flag
+        & info [ "all-contexts" ] ~doc:"Generate the lockdir for all contexts")
+    in
     let config = Common.init common in
     Scheduler.go ~common ~config (fun () ->
```

Comment:
You can put `@@` before `(fun` to get rid of some indentation (and then format).